[Docs] Add diagrams to circuit breaker (#1704)
App-vNext/Polly: docs/strategies/circuit-breaker.md

| `ManualControl`     | `null`                                                                     | Allows for manual control to isolate or close the circuit.                                 |
| `StateProvider`     | `null`                                                                     | Enables the retrieval of the current state of the circuit.                                 |

## Diagrams

### State diagram

```mermaid
%%{init: {'theme':'dark'}}%%
stateDiagram-v2
direction LR
    [*] --> Closed
    Closed --> Open: Exceeds threshold
    Open --> HalfOpen: Elapses break duration
    HalfOpen --> Closed: Passes the probe
    HalfOpen --> Open: Fails the probe
```

> Whenever someone says _the circuit breaks_ that means CB transitions from `Closed` to `Open` state.

Reviewer: ```suggestion
Whenever someone says _the circuit breaks_ that means the Circuit Breaker transitions from the `Closed` state to the `Open` state.
```

### Simple

> Let's suppose we have a circuit breaker strategy with `SamplingDuration`: `2 seconds`, `MinimumThroughput`: `2` and `FailureRatio` : `0.5`.

Reviewer: ```suggestion
Let's suppose we have a circuit breaker strategy with the following configuration:

- `SamplingDuration`: `2 seconds`;
- `MinimumThroughput`: `2`;
- `FailureRatio` : `0.5`.
```

#### Simple: happy path sequence diagram

The circuit will not break because the actual failure ratio (0.33) will be below the threshold (0.5) after the 3rd call.

```mermaid
%%{init: {'theme':'dark'}}%%
sequenceDiagram
    autonumber
    actor C as Caller
    participant P as Pipeline
    participant CB as CircuitBreaker
    participant D as DecoratedUserCallback

    C->>P: Calls ExecuteAsync
    P->>CB: Calls ExecuteCore
    Note over CB: Closed state

>     Note over CB, D: Sampling start

Reviewer: Might make sense to "activate" the line here to show the sampling being active in these diagrams.

    CB->>+D: Invokes
    D->>-CB: Returns result
    CB->>P: Returns result
    P->>C: Returns result

    C->>P: Calls ExecuteAsync
    P->>CB: Calls ExecuteCore
    CB->>+D: Invokes
    D->>-CB: Returns result
    CB->>P: Returns result
    P->>C: Returns result

    C->>P: Calls ExecuteAsync
    P->>CB: Calls ExecuteCore
    CB->>+D: Invokes
    D->>-CB: Fails
    Note over CB, D: Sampling end
    CB->>P: Propagates failure
    P->>C: Propagates failure
```

#### Simple: unhappy path sequence diagram

> The circuit will break because the actual failure ratio meets the threshold after the 2nd call.

Reviewer: ```suggestion
The circuit will break because the actual failure ratio meets the threshold (0.5) after the 2nd call.
```

```mermaid
%%{init: {'theme':'dark'}}%%
sequenceDiagram
    autonumber
    actor C as Caller
    participant P as Pipeline
    participant CB as CircuitBreaker
    participant D as DecoratedUserCallback

    C->>P: Calls ExecuteAsync
    P->>CB: Calls ExecuteCore
    Note over CB: Closed state
    Note over CB, D: Sampling start
    CB->>+D: Invokes
    D->>-CB: Returns result
    CB->>P: Returns result
    P->>C: Returns result

    C->>P: Calls ExecuteAsync
    P->>CB: Calls ExecuteCore
    CB->>+D: Invokes
    D->>-CB: Fails
    Note over CB: Moves to Open state
    CB->>P: Propagates failure
    P->>C: Propagates failure

    C->>P: Calls ExecuteAsync
    P->>CB: Calls ExecuteCore
    CB->>CB: Rejects request
    CB->>P: Throws <br/>BrokenCircuitException
    P->>C: Propagates exception
    Note over CB, D: Sampling end
```

### Complex

> Let's suppose we have a circuit breaker strategy with `SamplingDuration`: `2 seconds`, `MinimumThroughput`: `2`, `FailureRatio` : `0.5` and `BreakDuration`:`1 second`.

Reviewer: ```suggestion
Let's suppose we have a circuit breaker strategy with the following configuration:

- `SamplingDuration`: `2 seconds`;
- `MinimumThroughput`: `2`;
- `FailureRatio` : `0.5`;
- `BreakDuration`: `1 second`.
```
